src/header/common/origin.rs

```diff
@@ -0,0 +1,125 @@
+use header::{Header, Host, HeaderFormat};
+use std::fmt;
+use url::Url;
+use header::parsing::from_one_raw_str;
+
+/// The `Origin` header.
+///
+/// The `Origin` header is a version of the `Referer` header that is used for all HTTP fetches and `POST`s whose CORS flag is set.
+/// This header is often used to inform recipients of the security context of where the request was initiated.
+///
+///
+/// Following the spec, https://fetch.spec.whatwg.org/#origin-header, the value of this header is composed of
+/// a String (scheme), header::Host (host/port)
+///
+/// # Examples
+/// ```
+/// use hyper::header::{Headers, Origin, Host};
+///
+/// let mut headers = Headers::new();
+/// headers.set(
+///     Origin{
+///           scheme: "http".to_owned(),
+///           host: Host{hostname: "hyper.rs".to_owned(), port: None}
+///    }
+/// );
+/// ```
+/// ```
+/// use hyper::header::{Headers, Origin, Host};
+///
+/// let mut headers = Headers::new();
+/// headers.set(
+///     Origin{
+///           scheme: "https".to_owned(),
+///           host: Host{hostname: "wikipedia.org".to_owned(), port: Some(443)}
+///     }
+/// );
+/// ```
+
+#[derive(Clone, Debug)]
+pub struct Origin {
+    /// The scheme, such as http or https
+    pub scheme: String,
+    /// The host, such as Host{hostname: "hyper.rs".to_owned(), port: None}
+    pub host: Host,
+}
+
+impl Header for Origin {
+    fn header_name() -> &'static str {
+        static NAME: &'static str = "Origin";
+        NAME
+    }
+
+    fn parse_header(raw: &[Vec<u8>]) -> ::Result<Origin> {
+        from_one_raw_str(raw).and_then(|s: String| {
+            let url = Url::parse(&s).unwrap(); // unwrap so we can let url::ParseError Propogate
```

Comment:
Some discussion about parsing using `Url` has pointed out that we shouldn't accept an `Origin` that has a path part.

How about using this strategy instead:

1. Find `://`.
2. Scheme is `raw[..colon]`.
3. Then `Host::from_str(&raw[colon + 3..])`

I just took a peek, and saw there is no `FromStr` impl for `Host`, it'd probably be nice to move the `parse_header` code into a FromStr impl...